verilog/tools/ls/autoexpand.cc

```diff
@@ -1402,25 +1368,77 @@ std::optional<absl::string_view> AutoExpander::FindSpanToReplace(
     const TextStructureView &text_structure,
-    const std::vector<AutoExpander::Expansion> &expansions) {
-  std::vector<TextEdit> edits;
-  edits.reserve(expansions.size());
+    std::vector<AutoExpander::Expansion> expansions) {
+  std::sort(expansions.begin(), expansions.end(),
+            [](const AutoExpander::Expansion &first,
+               const AutoExpander::Expansion &second) {
+              return first.replaced_span.begin() < second.replaced_span.begin();
+            });
+
+  // To format expansions, we need to apply them to the source. Then we only
+  // call the formatter once, and produce a single TextEdit that replaces the
+  // entire file. This is orders of magnitude faster than formatting individual
+  // TextEdits.
+  std::string text;
+  auto begin = text_structure.Contents().begin();
+  for (const AutoExpander::Expansion &expansion : expansions) {
+    text.append(begin, expansion.replaced_span.begin());
+    text.append(expansion.new_text);
+    begin = expansion.replaced_span.end();
+  }
+  text.append(begin, text_structure.Contents().end());
+
+  VerilogAnalyzer analyzer(text, "");
+  const absl::Status analyze_status = analyzer.Analyze();
+  if (!analyze_status.ok()) {
+    LOG(ERROR) << "AUTO expansion produced invalid syntax. Aborting.";
+    return {};
+  }
+
+  FormatStyle format_style;
+  InitializeFromFlags(&format_style);
+
+  int64_t line_diff_acc = 0;
+  verible::LineNumberSet lines;
   for (const AutoExpander::Expansion &expansion : expansions) {
     const LineColumnRange range =
         text_structure.GetRangeForText(expansion.replaced_span);
-    edits.push_back(
-        {.range =
-             {
-                 .start = {.line = range.start.line,
-                           .character = range.start.column},
-                 .end = {.line = range.end.line, .character = range.end.column},
-             },
-         .newText = expansion.new_text});
+    // Formatter expects 1-indexed lines, hence the +1
+    const int start_line =
+        static_cast<int>(range.start.line + line_diff_acc + 1);
+    const int new_text_line_count =
+        std::count(expansion.new_text.begin(), expansion.new_text.end(), '\n') +
+        1;
+    const int end_line = static_cast<int>(start_line + new_text_line_count);
+    const int line_diff =
+        range.start.line + new_text_line_count - range.end.line - 1;
+    line_diff_acc += line_diff;
+    lines.Add({start_line, end_line});
   }
-  return edits;
+
+  std::string formatted;
+  const absl::Status format_status =
+      FormatVerilog(analyzer.Data(), "", format_style, &formatted, lines);
+
+  std::string &new_text = text;
+  if (format_status.ok()) {
+    new_text = formatted;
+  } else {
+    LOG(ERROR) << "Failed to format AUTO expanded code";
+  }
+
+  return {{.range =
+               {
+                   .start = {.line = 0, .character = 0},
+                   .end = {.line = (int)text_structure.Lines().size() - 1,
```

Comment:
To avoid fiddling with the lines and columns manually too much, you can also use the [GetRangeForText()](https://github.com/chipsalliance/verible/blob/56ec2f9938b39491ef0bee01d9718a19e88abe81/common/text/text_structure.h#L118-L120)  function to get the information, just passing it the full content.

(this is what happens in the formatter LSP call:)
https://github.com/chipsalliance/verible/blob/56ec2f9938b39491ef0bee01d9718a19e88abe81/verilog/tools/ls/verible-lsp-adapter.cc#L303-L310

Reply:
Great idea, thanks